libs/base/Control/Relation.idr

```diff
@@ -127,3 +127,59 @@ PartialEquivalence ty Equal where
 
 public export
 Equivalence ty Equal where
+
+----------------------------------------
+
+public export
+data SymClosure : Rel ty -> ty -> ty -> Type where
+  Fwd : {0 x, y : ty} -> rel x y -> SymClosure rel x y
+  Bwd : {0 x, y : ty} -> rel y x -> SymClosure rel x y
+
+public export
+Reflexive ty rel => Reflexive ty (SymClosure rel) where
+  reflexive = Fwd reflexive
+
+public export
+Symmetric ty (SymClosure rel) where
+  symmetric (Fwd xy) = Bwd xy
+  symmetric (Bwd yx) = Fwd yx
+
+----------------------------------------
+
+public export
+data TransClosure : Rel ty -> ty -> ty -> Type where
+  Nil : TransClosure rel x x
+  (::) : {x, y, z : ty} -> rel x y -> TransClosure rel y z -> TransClosure rel x z
```

Comment:
I think you only need `y : ty`, the others can be runtime irrelevant because they're determined by the endpoints which can be made runtime-relevant if need be.